### ISSUE TITLE
Help implementing Pinning images

### ISSUE THREAD

**Issue by TWiStErRob** [bumptech/glide#746]
Update the wiki or create a sample app that does this.

See:
 * https://github.com/bumptech/glide/issues/221
 * https://gist.github.com/Teovald/bfddf6cbddc8e61ade80
 * https://groups.google.com/forum/#!topic/glidelibrary/M2H-9EN-284
 * https://groups.google.com/forum/#!msg/glidelibrary/glDYWgKjNPw/U1mWSzTkCwAJ
 * https://github.com/Teovald/glide-wiki/blob/master/Caching-and-Cache-Invalidation.md#dialog-with-another-persistence-layer

**Comment by Teovald** [comment]
If you find it satisfactory, you can merge https://github.com/Teovald/glide-wiki/blob/master/Caching-and-Cache-Invalidation.md#dialog-with-another-persistence-layer  with Glide's wiki.  
I don't think I can make a pull request for the wiki, but you can just merge the diff.

**Comment by ersen-osman1** [comment]
Thanks for this

But regarding

https://gist.github.com/Teovald/bfddf6cbddc8e61ade80

I would like an example of how to  register it in a Glide Module and how to use it in a usual Glide request.

**Comment by Teovald** [comment]
you can use the samples provided by glide in order to get some examples.  
For example, the flickr sample shows you how to use a FlickrModelLoader.

**Comment by ersen-osman1** [comment]
Hi Thanks but this is different from the two classes you have provided in the gist above. I want an example on how to use the ASynchronizableDataFetcher and ASynchronizableImageLoader because the description listed in wiki matches my requirements exactly.

**Comment by TWiStErRob** [comment]
He pointed to those samples because all model loaders needs to be registered the same. Here's a rundown:
 * extend both classes found in gist
 * register the new Loader via a [`GlideModule`](https://github.com/bumptech/glide/wiki/Configuration#creating-a-glidemodule)
 * implement abstract methods in the new classes
  * the new Loader needs to return the new Fetcher
  * the new Fetcher queries your pinned storage based on what's in model

**Comment by ersen-osman** [comment]
•extend both classes found in gist

Hello can you explain this again?, did you mean implement?

**Comment by ersen-osman** [comment]
```
public class PieceLoader extends ASynchronizableDataFetcher<Piece>{

    public PieceLoader(@Nullable DataFetcher<InputStream> networkFetcher, @NonNull Piece piece, int width, int height) {
        super(networkFetcher, piece, width, height);
    }

    @Nullable
    @Override
    public InputStream loadFromSynchro(@NonNull Piece piece, int width, int height) {
        return null;
    }

    @Override
    public String getId() {
        return null;
    }
}
```
This is what I have so far, I am confused with what I do with ASynchronizableDataFetcher and ASynchronizableDataFetcher

**Comment by TWiStErRob** [comment]
Exactly, that's what I meant by extending. Do the same for the other ASync... class and implement the missing methods as you want.

**Comment by ersen-osman** [comment]
Thanks. So I make another class called PieceFetcher which will extend ASynchronizableImageLoader?

**Comment by TWiStErRob** [comment]
Yes, but try to be a little more consistent with naming... `...Fetcher extends A...Fetcher` and same for Loader.

**Comment by ersen-osman** [comment]
Thanks, I got them mixed up :) .

**Comment by ersen-osman** [comment]
Is there a sample I can refer too? I do not know what to next :S. The Flickr sample seems different with these two classes. My model already contains URLs and absolute file paths to the image content already.

**Comment by TWiStErRob** [comment]
I wish I could point to well-rounded sample, that's why this issue exists, to make some clear and easily adaptable sample/docs.

**Comment by ersen-osman** [comment]
Thanks. I will play around with it and post my findings, if all goes well I can make a sample :)

**Comment by Teovald** [comment]
IMO, the wiki addition is enough since the rest of the stack is explained either on the wiki or in the samples but I wrote the code so I am obviously biased.  

Could you please elaborate on what you are missing Ersen ?

I don't mind writing a 'full' sample if it helps (excluding the implementation of the separate 'pinning' module because that's really up to you).

**Comment by erseno** [comment]
Hi guys, I hope you had a great Christmas and Happy New Year. I apologise for the late reply. I think a full sample would be great so I decided to start one here to use as a practical example / starting point to demonstrate the usage. 

https://github.com/erseno/Glide-Pinning-Image-Sample

However, since I am still trying to find out how it works, I have not written any model loader yet. I prepared a simple app which covers all the mundane things :P 

Teovald, I would be extremely grateful if you could fill in the missing implementation and send a pull request.

**Comment by Teovald** [comment]
Hi, 

happy new year to you too.  

I will try to have a look asap, but probably not until this weekend.

**Comment by erseno** [comment]
Many thanks!

**Comment by Teovald** [comment]
done.
have a look at : https://github.com/Teovald/Glide-Pinning-Image-Sample

**Comment by Teovald** [comment]
re-reading the wiki article I wrote, I will also clean it up a little bit once I find the time to do so.

**Comment by Teovald** [comment]
done. it should be better now.

**Comment by erseno** [comment]
Hi Teovald, apologies for the late reply again. Thank you for your contribution :+1:

**Comment by Teovald** [comment]
You are welcome.  

Between the official glide wiki, my additions and the code sample you should have enough information to figure out how this works.  
Still, let me know what you think of the doc & the sample.  
FWIIW, we have been using similar ModelLoaders/Fetchers on production with several millions daily active users for almost a year, so I am pretty sure that it is solid. :)
